qemu/tests/netperf_udp_perf.py

```diff
@@ -0,0 +1,229 @@
+from virttest import error_context
+from provider import netperf_base
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    Netperf UDP_STREAM test with netperf.
+
+    1) Boot up VM
+    2) Prepare the test environment in server/client/host
+    3) Run netserver on guest, run netperf client on remote host
+    4) Collect results
+
+    :param test: QEMU test object.
+    :param params: Dictionary with the test parameters.
+    :param env: Dictionary with test environment.
+    """
+
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+    login_timeout = int(params.get("login_timeout", 360))
+
+    try:
+        vm.wait_for_serial_login(timeout=login_timeout, restart_network=True).close()
+    except virt_vm.VMIPAddressMissingError:
+        pass
+
+    mac = vm.get_mac_address(0)
+    server_ip = vm.wait_for_get_address(0, timeout=90)
+
+    if len(params.get("nics", "").split()) > 1:
+        server_ctl = vm.wait_for_login(nic_index=1, timeout=login_timeout)
+        server_ctl_ip = vm.wait_for_get_address(1, timeout=90)
+    else:
+        server_ctl = vm.wait_for_login(timeout=login_timeout)
+        server_ctl_ip = server_ip
+
+    logging.debug(process.system_output("numactl --hardware",
+                                        verbose=False, ignore_status=True,
+                                        shell=True).decode())
+    logging.debug(process.system_output("numactl --show",
+                                        verbose=False, ignore_status=True,
+                                        shell=True).decode())
+    # pin guest vcpus/memory/vhost threads to last numa node of host by default
+    numa_node = netperf_base.pin_vm_threads(vm, params.get("numa_node"))
+    host = params.get("host", "localhost")
+    host_ip = host
+    client = params.get("client", "localhost")
+    client_ip = client
+    client_pub_ip = params.get("client_pub_ip")
+    client = remote.wait_for_login(params.get("shell_client_client"),
+                                   client_pub_ip,
+                                   params.get("shell_port_client"),
+                                   params.get("username_client"),
+                                   params.get("password_client"),
+                                   params.get("shell_prompt_client"))
+    cmd = "ifconfig %s %s up" % (params.get("client_physical_nic"),
+                                 client_ip)
+    netperf_base.ssh_cmd(client, cmd)
+
+    error_context.context("Prepare env of server/client/host", logging.info)
+    prepare_list = set([server_ctl, client, host])
+    tag_dict = {server_ctl: "server", client: "client", host: "host"}
+    ip_dict = {server_ctl: server_ctl_ip, client: client_pub_ip,
+               host: host_ip}
+    for i in prepare_list:
+        params_tmp = params.object_params(tag_dict[i])
+        shell_port = int(params_tmp["shell_port"])
+        password = params_tmp["password"]
+        username = params_tmp["username"]
```

Comment:
These lines can be moved into the function since you pass params to it.

Reply:
updated.